odbc/src/api/data.rs

```diff
@@ -140,6 +145,7 @@ impl IntoCData for Bson {
 
 pub unsafe fn format_and_return_bson(
```

Comment:
From what I can see, the code is always going to retrieve the data and then it will try to convert it.
I think it is more efficient to check if the conversion is valid first. No need to get the data if we know of the bat that we will ditch it because the conversion to the target C type is not supported.
Also, with the introduction of the var_data_indicies data structure, I think we could also before even asking to get the data, stop if the data for the column has already been retrieved entirely. Maybe this means changing the offset to be 0 = not retrieved yet and 1 = already retrieved for fixed size data.
These are maybe small optimizations and if the application using the driver isn't too dumb, they should not call SQLGetData on fixed length data again, but if this is not the code and they always call SQLGetData until they get NO_DATA then we will waste time and if the resulset is big, it can add up and make a real difference.